README.md
![ ERM19264 ](https://github.com/gavinlyonsrepo/ERM19264_UC1609/blob/main/extras/image/color.jpg)

Table of contents
---------------------------

  * [Overview](#overview)
  * [Output](#output)
  * [Installation](#installation)
  * [Hardware](#hardware)
  * [Features](#features)
  * [Files](#files)
  * [Ports](#ports)
  
Overview
--------------------
* Name : ERM19264_UC1609
* Title : Library for ERM19264-5 v3 LCD  (UC1609C controller) for the Arduino eco-system
* Description : 

1. Arduino library.      
2. Inverse, rotate and contrast control. 
3. ASCII text strings and character text display.
4. Graphics library included based on the Adafruit GFX library.
5. Sleep mode.
6. 3 different modes: Multi-buffer , single buffer , light weight text only
7. custom bitmaps supported.
 
* Author: Gavin Lyons
* Arduino IDE: 1.8.10

* Tested on following MCUs both software and hardware SPI,
The example files are setup for an UNO for the pin connections used 
by for other MCU testing see extras folder GPIO_MCU_used.txt file.
1. Arduino  UNO & NANO v3
2. ESP8266 
3. ESP32 
4. STM32 "blue pill"

Output
---------------------------------

Output Screenshots, From left to right top to bottom.

1. Full screen bitmap displayed
2. Multi buffer mode screen divided into two buffers
3. Different size and type of fonts 
4. Available ASCII font printed out, this can be expanded to [extended ASCII font](https://www.extended-ascii.com/) by minor modification to file see features.

![op](https://github.com/gavinlyonsrepo/ERM19264_UC1609/blob/main/extras/image/output.jpg)


Installation
------------------------------

The library is included in the official Arduino library manger and the optimum way to install it is using the library manager which can be opened by the *manage libraries* option in Arduino IDE. 

See link below for instruction for this and for the other methods too.

[Installing Additional Arduino Libraries guide](https://www.arduino.cc/en/Guide/Libraries)

Hardware
----------------------------

9 pins , Vcc and GND, anode and cathode for the backlight LED and an SPI interface.
The example files are setup for an UNO for the pin connections used by for other MCU tested see extras folder GPIO_MCU_used.txt file. The backlight control is left up to user.
If  using Hardware SPI two of  pins will be tied to the SPI CLK and MOSI lines if using software SPI you should be able use any GPIO you want for all five pins.
Datasheets are in the extras folder. 

There are 3 different colours in range, Parts used purchased from [ebay](https://www.ebay.ie/itm/2-inch-White-192x64-Graphic-LCD-Display-Module-UC1609-SPI-for-Arduino/293617684779?hash=item445cfa512b:g:10MAAOSwYV9e6xsi)
 
1. ERM19264SBS-5 V3 LCD Display UC1609C controller ,  white on blue
2. ERM19264FS-5 V3 LCD Display  UC1609C controller , black on white
3. ERM19264DNS-5 V3 LCD Display  UC1609C controller white on black

The library was tested on 1 and 2. 
The UC1609 controller chip is a 3.3V device but the ERM LCD module has a "662k" 3.3V regulator at back.
So the ERM LCD module will run at 5V as well if this is present.
It was always run it at 3.3V during testing. 
The Backlight should always be connected to 3.3V according to datasheets.  

This wiring Diagram from the manufacturer showing hardware setup connected to an ~8051 MCU.
Showing both 5 volt and 3.3 volt systems.

![ ERM19264 ](https://github.com/gavinlyonsrepo/ERM19264_UC1609/blob/main/extras/image/connect.jpg)

Features
-------------------------

*SPI*

Hardware and software SPI. Two different class constructors. User can pick the relevant constructor, see examples files. Hardware SPI is recommended, far faster and more reliable but Software SPI allows for more flexible GPIO selection and easier to port to other MCU' s. When running Software SPI it may be necessary on very high frequency MCU's to change the UC1609_HIGHFREQ_DELAY define, It is a microsecond delay by default it is at 0. All the hardware SPI settings are defined in the header file and can be changed if necessary. 

*buffers*

3 buffer modes 

1. MULTI_BUFFER (default)
2. SINGLE_BUFFER 
3. NO_BUFFER , Text only no buffer , relatively light weight. A "hello world" Sketch uses 2320 bytes (7%) of and 42 bytes (2%) of dynamic memory. Turns LCD into simple character LCD(216 characters)

To switch between modes, user must make a change to the USER BUFFER OPTION SECTION  at top of 
ERM19264_UC1609.h file.  Pick ONE option and one option ONLY. The example files at top, say which option to pick. If wrong option is picked, example files will not work or maybe even compile.
Bitmaps can still be written directly to screen in NO_BUFFER mode but no graphics possible.

*fonts*

The ASCII font(custom_font in the custom_graphics_font.h file) is truncated by a define ( UC_FONT_MOD_TWO) after first 127 characters (see output pic) to save memory space(640 bytes), if you wish to use rest of the [extended ASCII font](https://www.extended-ascii.com/), simply comment this define out. Extended ASCII will not work in no buffer mode (text only) for v1.1.0 see TODO_ISSUES.TXT
file in extras for solutions. 
The font is a standard 5 by 7 ASCII font with two  columns  of padding added. So 7 by 8 in effect. In standard text size and "no buffer" mode, this means: 192/7 * 64/8 = 27 * 8 = 216 characters. 

*User adjustments*

When the user calls LCDbegin() to start LCD they can specify a contrast setting from 0x00 to 0xFF.
Datasheet says 0x49 is default. (VbiasPOT)

It is also possible for user to change LCD bias ,  Temperature coefficient, frame rate and power control but this must be done by changing defines in header file. Choose lower frame rate for lower power, and choose higher frame rate to improve LCD contrast and minimize flicker. See Data sheet for range of values
here. Defaults where found to be fine during all testing of this library.


| Parameter | default Values |  Define | Register |
| ------ | ------ |  ------ | ------ |
| LCD bias |  9 | BIAS_RATIO_SET | BR 1:0 |
| Temp coefficient | -0.00%/ C |  TEMP_COMP_SET | TC 1:0  |
| Frame rate | 95 fps |  FRAMERATE_SET |  LC 4:3 |
| Power control | 1.4mA + internal V LCD |  PC_SET | PC 2:0 |
| V bias Bot(contrast) | 0x49h default|  Set by user with LCDbegin | PM 7:0 |

*Functions*

Functions: Detailed information on the functions can be found in comments in the ERM19264_UC1609.cpp  file and a list of them in keywords.txt. The graphic functions can be found in the custom_graphic.h file. 

Files
-------------------

| Src Files| Desc |
| ------ | ------ |
| ERM19264_UC1609.h | library header file  |
| ERM19264_UC1609.cpp |  library  source file  |
| custom_graphics.h | Custom graphics header file |
| custom_graphics.cpp | Custom graphics source file |
| custom_graphics_font.h | Custom graphics  font  file |

| Examples files ino  | Desc |
| ------ | ------ |
| BITMAP | Shows use of bitmaps  |
| GRAPHICS |  Shows use of graphics   |
| MISC | Shows misc functions, rotate invert etc |
| MULTIBUFFER | Shows use of multi buffer mode |
| NOBUFFER | Shows use of no buffer text only mode |
| TEXT | Shows use of text IN buffer mode   |
| SINGLEBUFFER| Shows use of single buffer mode |
| SWSPI | Shows use of software SPI |

Ports
------------------------------------------

* ERM19264_UC1609_T (T for text). Light weight Text only version for arduino ecosystem [here at link](https://github.com/gavinlyonsrepo/ERM19264_UC1609_T)

* PIC xc8 ports [Link](https://github.com/gavinlyonsrepo/pic_16F18346_projects)



extras/CHANGELOG.md
# Changelog

* version 1.0.0 October 2020
	* first release

* version 1.1.0 April 2021
	* Minor update 
	* Typos in Keyword.txt and,  Example file -MISC  removed.
	* Delay <UC1609_RESET_DELAY2> reduced, asERM19264SBS (white on blue)version of LCD on  Software SPI would not initialise, intermittently.
	* Example -TEST new test added to print out extended ASCII font.


src/ERM19264_UC1609.h
/*
* Project Name: ERM19264_UC1609
* File: ERM19264_UC1609.h
* Description: ERM19264 LCD driven by UC1609C controller header file
* Author: Gavin Lyons.
* URL: https://github.com/gavinlyonsrepo/ERM19264_UC1609
*/

#ifndef ERM19264_UC1609_H
#define ERM19264_UC1609_H

#if (ARDUINO >=100)
  #include "Arduino.h"
#else
  #include "WProgram.h"
#endif


//***********************************************
// ** USER BUFFER OPTION SECTION **********
// Pick one buffer option and ONE option ONLY 
#define MULTI_BUFFER // (1) Note default
//#define SINGLE_BUFFER // (2) 
//#define NO_BUFFER   // (3) No graphics, text + bitmap only
// **********************************************
// **********************************************


#ifdef NO_BUFFER
   #include "custom_graphics_font.h"
#else
   #include "custom_graphics.h"
#endif

#include <SPI.h>

// Display Pixel colours   definition
// (1): white on blue,  FG = white BG = blue
// ERM19264SBS-4 LCD Display White on Blue
// (2): black on white, FG = black BG = white
// ERM19264FS-4 LCD Display Black on White
// (3):white on black,  FG = white, BG = black
// ERM19264DNS-4LCD Display White on Black
#define FOREGROUND  0
#define BACKGROUND 1
#define INVERSE 2

// UC1909 Read registers
#define UC1609_GET_STATUS 0xFE

// UC1909 Write registers
#define UC1609_SYSTEM_RESET 0xE2 // Not used v1.x

#define UC1609_POWER_CONTROL 0x2F 
#define UC1609_PC_SET 0x06 // PC[2:0] 110 Internal V LCD (7x charge pump) + 10b: 1.4mA

#define UC1609_ADDRESS_CONTROL 0x88 // set RAM address control
#define UC1609_ADDRESS_SET 0x02 // Set AC [2:0] Program registers  for RAM address control.

#define UC1609_SET_PAGEADD 0xB0 // Page address Set PA[3:0]
#define UC1609_SET_COLADD_LSB 0x00 // Column Address Set CA [3:0]
#define UC1609_SET_COLADD_MSB 0x10 // Column Address Set CA [7:4]

#define UC1609_TEMP_COMP_REG 0x27 // Temperature Compensation Register
#define UC1609_TEMP_COMP_SET 0x00 // TC[1:0] = 00b= -0.00%/ C

#define UC1609_FRAMERATE_REG 0xA0 // Frame rate
#define UC1609_FRAMERATE_SET 0x01  // Set Frame Rate LC [4:3] 01b: 95 fps

#define UC1609_BIAS_RATIO 0xE8 // Bias Ratio. The ratio between V-LCD and V-D .
#define UC1609_BIAS_RATIO_SET 0x03 //  Set BR[1:0] = 11 (set to 9 default)

#define UC1609_GN_PM 0x81 // Set V BIAS Potentiometer to fine tune V-D and V-LCD  (double-byte command)
#define UC1609_DEFAULT_GN_PM 0x49 // default only used if user does not specify Vbias

#define UC1609_LCD_CONTROL 0xC0 // Rotate map control
#define UC1609_DISPLAY_ON 0xAE // enables display
#define UC1609_ALL_PIXEL_ON 0xA4 // sets on all Pixels on
#define UC1609_INVERSE_DISPLAY 0xA6 // inverts display
#define UC1609_SCROLL 0x40 // scrolls , Set the scroll line number. 0-64

// Rotate
#define UC1609_ROTATION_FLIP_TWO 0x06
#define UC1609_ROTATION_NORMAL 0x04
#define UC1609_ROTATION_FLIP_ONE 0x02
#define UC1609_ROTATION_FLIP_THREE 0x00

// Delays
#define UC1609_RESET_DELAY 50 // mS delay
#define UC1609_RESET_DELAY2   0 // mS delay
#define UC1609_INIT_DELAY 100   //  mS delay
#define UC1609_INIT_DELAY2 3 // mS delay

// No buffer mode font
#define UC1609_ASCII_OFFSET 0x20 //0x20, ASCII character for Space
#define UC1609_FONTPADDING  send_data(0x00)
#define UC1609_FONTWIDTH 5

// GPIO
#define UC1609_CS_SetHigh digitalWrite(_LCD_CS, HIGH)
#define UC1609_CS_SetLow digitalWrite(_LCD_CS, LOW)
#define UC1609_CD_SetHigh digitalWrite(_LCD_CD, HIGH)
#define UC1609_CD_SetLow digitalWrite(_LCD_CD, LOW)
#define UC1609_RST_SetHigh digitalWrite(_LCD_RST, HIGH)
#define UC1609_RST_SetLow digitalWrite(_LCD_RST, LOW)

// SPI
#define SPI_FREQ 8000000 // Mhz (1000000 can be used on high freq MCU)
#define SPI_DIRECTION  MSBFIRST 
#define SPI_UC1609_MODE SPI_MODE0
#define SPI_UC1609_CLOCK_DIV SPI_CLOCK_DIV8 //STM32 uses this 
#define UC1609_HIGHFREQ_DELAY 0 // Can be used in software SPI for high freq MCU

// Display  Size
const uint8_t LCD_WIDTH = 192;
const uint8_t LCD_HEIGHT = 64;

#ifdef MULTI_BUFFER

struct MultiBuffer
{
  uint8_t* screenbitmap; // pointer to buffer
  uint8_t width=LCD_WIDTH;  // bitmap x size
  uint8_t height=LCD_HEIGHT ; // bitmap y size
  int16_t xoffset = 0; // x offset
  int16_t yoffset = 0; // y offset
  int8_t data = 0;  
};
#endif

//class 
#ifdef NO_BUFFER
class ERM19264_UC1609 {
#else
class ERM19264_UC1609 : public custom_graphics {
#endif
  public:
     // Contructor 1 Software SPI with explicit SCLK and SDIN
    ERM19264_UC1609(int8_t cd, int8_t rst, int8_t cs, int8_t sclk, int8_t din);
    // Constructor 2 Hardware SPI
    ERM19264_UC1609(int8_t cd, int8_t rst, int8_t cs);

    ~ERM19264_UC1609(){};

#ifdef MULTI_BUFFER
   MultiBuffer* ActiveBuffer;
#endif

#ifdef SINGLE_BUFFER
    uint8_t* buffer;
    uint8_t bufferWidth = LCD_WIDTH;
    uint8_t bufferHeight = LCD_HEIGHT;
#endif

// Functions not needed for no_buffer mode 
#ifndef NO_BUFFER
    virtual void drawPixel(int16_t x, int16_t y, uint16_t colour) override;
     void LCDupdate(void);
     void LCDclearBuffer(void);
     void LCDBuffer(int16_t x, int16_t y, uint8_t w, uint8_t h, uint8_t* data);
#else  // Functions only needed for one of  buffer mode's
     void  LCDNoBufferGotoXY(uint8_t column , uint8_t page);
     void  LCDNoBufferChar(unsigned char character);
     void  LCDNoBufferString(const unsigned char *characters);
#endif
    
    void LCDbegin(uint8_t VbiasPot = UC1609_DEFAULT_GN_PM );
    void LCDinit(void);
    void LCDEnable(uint8_t on);
    void LCDFillScreen(uint8_t pixel, uint8_t mircodelay);
    void LCDFillPage(uint8_t pixels);
    void LCDrotate(uint8_t rotatevalue);
    void invertDisplay(uint8_t on);
    void LCD_allpixelsOn(uint8_t bits);
    void LCDscroll(uint8_t bits);
    void LCDReset(void);
    void LCDBitmap(int16_t x, int16_t y, uint8_t w, uint8_t h, const uint8_t* data);
    uint16_t LCDreadStatus();
           
  private:

    void send_data(uint8_t data);
    void send_command(uint8_t command, uint8_t value);
    bool isHardwareSPI(void);
    void CustomshiftOut(uint8_t bitOrder, uint8_t val);

    int8_t _LCD_CS;
    int8_t _LCD_CD;
    int8_t _LCD_RST;
    int8_t _LCD_SCLK; // Software SPI only
    int8_t _LCD_DIN;  // Software SPI only
    uint8_t _VbiasPOT; // Contrast default 0x49 datasheet 00-FE
};

#endif
